src/transforms/apply-operation.js

```diff
@@ -120,14 +132,46 @@ function insertText(state, operation) {
  *
  * @param {State} state
  * @param {Object} operation
+ *   @param {Boolean} operation.deep (optional) Join recursively the
+ *   respective last node and first node of the nodes' children. Like a zipper :)
  * @return {State}
  */
 
 function joinNode(state, operation) {
-  const { path, withPath } = operation
-  let { document } = state
-  document = document.joinNode(path, withPath)
-  state = state.merge({ document })
+  const { path, withPath, deep = false } = operation
```

Comment:
Continuing the idea from above, why not always have `joinNode` be deep and "zipper" the nodes together?

Or, if that doesn't work... what about having `joinNode` and `splitNode` take a `depth` argument, which defaults to `Infinity`, but can be passed in for cases where we need the control instead?